File: lib/services/encryption_service.dart
import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';
import 'package:crypto/crypto.dart';
import 'package:pointycastle/export.dart';

/// End-to-end encryption service.
///
/// Flow:
///   passphrase → SHA-256 → record ID (used to identify the row in Supabase)
///   passphrase → PBKDF2(SHA-256, 100000 iterations) → 32-byte AES key
///   AES-256-GCM(key, random 12-byte IV) → encrypted payload (base64)
class EncryptionService {
  static const _pbkdf2Salt = 'chulemav1'; // app-specific salt, not secret
  static const _pbkdf2Iterations = 100000;
  static const _keyLength = 32; // AES-256
  static const _ivLength = 12;  // GCM standard

  // ── Public API ─────────────────────────────────────────────────────────────

  /// Encrypts raw [data] bytes using AES-256-GCM.
  /// Returns bytes: IV (12) + ciphertext + auth tag.
  static Uint8List encryptBytes(String passphrase, Uint8List data) {
    final key = _deriveKey(passphrase);
    final iv  = _randomBytes(_ivLength);

    final cipher = GCMBlockCipher(AESEngine())
      ..init(
        true,
        AEADParameters(KeyParameter(key), 128, iv, Uint8List(0)),
      );

    final output   = cipher.process(data);
    final combined = Uint8List(_ivLength + output.length);
    combined.setRange(0, _ivLength, iv);
    combined.setRange(_ivLength, combined.length, output);
    return combined;
  }

  /// Decrypts bytes produced by [encryptBytes].
  static Uint8List decryptBytes(String passphrase, Uint8List combined) {
    final key        = _deriveKey(passphrase);
    final iv         = combined.sublist(0, _ivLength);
    final ciphertext = combined.sublist(_ivLength);

    final cipher = GCMBlockCipher(AESEngine())
      ..init(
        false,
        AEADParameters(KeyParameter(key), 128, iv, Uint8List(0)),
      );

    return cipher.process(Uint8List.fromList(ciphertext));
  }

  /// Derives the Supabase record ID from the passphrase.
  /// This is SHA-256(passphrase) as a hex string.
  static String deriveRecordId(String passphrase) {
    final bytes = utf8.encode(passphrase.trim());
    return sha256.convert(bytes).toString();
  }

  /// Encrypts [plaintext] using AES-256-GCM with a key derived from [passphrase].
  /// Returns a base64-encoded string: base64(IV + ciphertext + authTag).
  static String encrypt(String passphrase, String plaintext) {
    final key = _deriveKey(passphrase);
    final iv  = _randomBytes(_ivLength);

    final cipher = GCMBlockCipher(AESEngine())
      ..init(
        true, // encrypt
        AEADParameters(KeyParameter(key), 128, iv, Uint8List(0)),
      );

    final input  = Uint8List.fromList(utf8.encode(plaintext));
    final output = cipher.process(input);

    // Prepend IV so we can recover it during decryption
    final combined = Uint8List(_ivLength + output.length);
    combined.setRange(0, _ivLength, iv);
    combined.setRange(_ivLength, combined.length, output);

    return base64.encode(combined);
  }

  /// Decrypts a payload produced by [encrypt].
  /// Returns the original plaintext string, or throws on bad passphrase / tampered data.
  static String decrypt(String passphrase, String encryptedBase64) {
    final key      = _deriveKey(passphrase);
    final combined = base64.decode(encryptedBase64);

    final iv         = combined.sublist(0, _ivLength);
    final ciphertext = combined.sublist(_ivLength);

    final cipher = GCMBlockCipher(AESEngine())
      ..init(
        false, // decrypt
        AEADParameters(KeyParameter(key), 128, iv, Uint8List(0)),
      );

    final decrypted = cipher.process(Uint8List.fromList(ciphertext));
    return utf8.decode(decrypted);
  }

  // ── Private helpers ────────────────────────────────────────────────────────

  static Uint8List _deriveKey(String passphrase) {
    final saltBytes  = utf8.encode(_pbkdf2Salt);
    final passBytes  = utf8.encode(passphrase.trim());

    final pbkdf2 = PBKDF2KeyDerivator(HMac(SHA256Digest(), 64))
      ..init(Pbkdf2Parameters(
        Uint8List.fromList(saltBytes),
        _pbkdf2Iterations,
        _keyLength,
      ));

    return pbkdf2.process(Uint8List.fromList(passBytes));
  }

  static Uint8List _randomBytes(int length) {
    final rng   = Random.secure();
    final bytes = Uint8List(length);
    for (var i = 0; i < length; i++) {
      bytes[i] = rng.nextInt(256);
    }
    return bytes;
  }
}

File: ios/Runner/Sources/Services/EncryptionService.swift
import Foundation
import CryptoKit
import CommonCrypto

/// Matches Flutter's EncryptionService exactly:
///   - Record ID:  SHA-256(passphrase.trim()) → lowercase hex
///   - Key:        PBKDF2-SHA256(passphrase.trim(), salt="chulemav1", iter=100000, keyLen=32)
///   - Payload:    base64( IV(12) + AES-256-GCM(ciphertext + 16-byte tag) )
enum EncryptionService {
    private static let pbkdf2Salt       = "chulemav1"
    private static let pbkdf2Iterations = 100_000
    private static let keyLength        = 32  // AES-256
    private static let ivLength         = 12  // GCM nonce

    // MARK: - Public API

    static func deriveRecordId(_ passphrase: String) -> String {
        let data   = Data(passphrase.trimmingCharacters(in: .whitespaces).utf8)
        let digest = SHA256.hash(data: data)
        return digest.map { String(format: "%02x", $0) }.joined()
    }

    /// Encrypts plaintext string → base64(IV + ciphertext + tag)
    static func encrypt(_ passphrase: String, _ plaintext: String) throws -> String {
        let key  = try deriveKey(passphrase)
        let data = Data(plaintext.utf8)
        let encrypted = try encryptBytes(key: key, data: data)
        return encrypted.base64EncodedString()
    }

    /// Decrypts base64(IV + ciphertext + tag) → plaintext string
    static func decrypt(_ passphrase: String, _ encryptedBase64: String) throws -> String {
        guard let combined = Data(base64Encoded: encryptedBase64) else {
            throw EncryptionError.invalidBase64
        }
        let key       = try deriveKey(passphrase)
        let decrypted = try decryptBytes(key: key, combined: combined)
        guard let plaintext = String(data: decrypted, encoding: .utf8) else {
            throw EncryptionError.invalidUTF8
        }
        return plaintext
    }

    /// Encrypts raw Data → Data(IV + ciphertext + tag)
    static func encryptBytes(_ passphrase: String, _ data: Data) throws -> Data {
        let key = try deriveKey(passphrase)
        return try encryptBytes(key: key, data: data)
    }

    /// Decrypts Data(IV + ciphertext + tag) → raw Data
    static func decryptBytes(_ passphrase: String, _ combined: Data) throws -> Data {
        let key = try deriveKey(passphrase)
        return try decryptBytes(key: key, combined: combined)
    }

    // MARK: - Private

    private static func deriveKey(_ passphrase: String) throws -> SymmetricKey {
        let passData = Data(passphrase.trimmingCharacters(in: .whitespaces).utf8)
        let saltData = Data(pbkdf2Salt.utf8)

        var derivedKey = [UInt8](repeating: 0, count: keyLength)
        let result = passData.withUnsafeBytes { passPtr in
            saltData.withUnsafeBytes { saltPtr in
                CCKeyDerivationPBKDF(
                    CCPBKDFAlgorithm(kCCPBKDF2),
                    passPtr.baseAddress, passData.count,
                    saltPtr.baseAddress, saltData.count,
                    CCPseudoRandomAlgorithm(kCCPRFHmacAlgSHA256),
                    UInt32(pbkdf2Iterations),
                    &derivedKey, keyLength
                )
            }
        }
        guard result == kCCSuccess else { throw EncryptionError.keyDerivationFailed }
        return SymmetricKey(data: Data(derivedKey))
    }

    private static func encryptBytes(key: SymmetricKey, data: Data) throws -> Data {
        let nonce  = AES.GCM.Nonce()
        let sealed = try AES.GCM.seal(data, using: key, nonce: nonce)

        // Layout: IV(12) + ciphertext + tag(16) — matches Flutter/PointyCastle output
        var combined = Data()
        combined.append(contentsOf: nonce)
        combined.append(sealed.ciphertext)
        combined.append(sealed.tag)
        return combined
    }

    private static func decryptBytes(key: SymmetricKey, combined: Data) throws -> Data {
        guard combined.count > ivLength + 16 else { throw EncryptionError.tooShort }

        let nonce      = try AES.GCM.Nonce(data: combined.prefix(ivLength))
        let tagStart   = combined.count - 16
        let ciphertext = combined[ivLength..<tagStart]
        let tag        = combined[tagStart...]

        let sealed = try AES.GCM.SealedBox(nonce: nonce, ciphertext: ciphertext, tag: tag)
        return try AES.GCM.open(sealed, using: key)
    }
}

enum EncryptionError: Error {
    case keyDerivationFailed
    case invalidBase64
    case invalidUTF8
    case tooShort
}
